stdlib/json.gr

```diff
@@ -0,0 +1,2155 @@
+              CompactObjectEntries => {
+                Buffer.addChar(':', buffer)
+              },
+              OneObjectEntryPerLine => {
+                Buffer.addChar(':', buffer)
+                Buffer.addChar(' ', buffer)
+              },
+            }
+
+            let err = printElementPretty(value, implHelper, elemLevel)
+            if (Option.isSome(err)) return err
+
+            match (currentRest) {
+              [] => break,
+              [newHead, ...newRest] => {
+                currentHead = newHead
+                currentRest = newRest
+              },
+            }
+          }
+
+          if (format.objectFormat == OneObjectEntryPerLine) {
+            implHelper.printNewLine()
+            implHelper.printIndentation(indentationLevel)
+          }
+
+          Buffer.addChar('}', buffer)
+
+          return None
+        },
+      }
+    },
+  }
+}
+
+let isCodePointInBasicMultilingualPlane = (code: Number) =>
+  code >= 0x0000 &&
+    code <= 0xFFFF
+
+let isHighSurrogate = (code: Number) => code >= 0xD800 && code <= 0xDBFF
+
+let isLowSurrogate = (code: Number) => code >= 0xDC00 && code <= 0xDFFF
+
+let combineSurrogatePairToCodePoint =
+  (
+    highSurrogate: Number,
+    lowSurrogate: Number,
+  ) => {
+  // If this was a method exposed by itself in a library then it should check the
+  // ranges of the input surrogates, but here it's necessary because checks are made
+  // as part of the parsing logic.
+  ((highSurrogate - 0xD800) << 10) + (lowSurrogate - 0xDC00) + 0x10000
+}
+
+let makePrettyPrintJsonWriter =
+  (
+    format: FormattingSettings,
+    buffer: Buffer.Buffer,
+  ) => {
+  let printNewLine = match (format.lineEnding) {
+    NoLineEnding => () => void,
+    LineFeed =>
+      () => {
+        Buffer.addChar('\n', buffer)
+      },
+    CarriageReturnLineFeed =>
+      () => {
+        Buffer.addChar('\r', buffer)
+        Buffer.addChar('\n', buffer)
+      },
+    CarriageReturn =>
+      () => {
+        Buffer.addChar('\r', buffer)
+      },
+  }
+
+  let printIndentation = match (format.indentation) {
+    IndentWithTab =>
+      indentationLevel => {
+        let mut count = 0
+        while (count < indentationLevel) {
+          Buffer.addChar('\t', buffer)
+          count += 1
+        }
+      },
+    IndentWithSpaces(spacesPerIndentation) =>
+      indentationLevel => {
+        let mut count = 0
+        let spaceCount = indentationLevel * spacesPerIndentation
+        while (count < spaceCount) {
+          Buffer.addChar(' ', buffer)
+          count += 1
+        }
+      },
+    NoIndentation => indentationLevel => void,
+  }
+
+  // A possible optimization to make this faster would be to
+  // prepare a different closure for each combination fo escaping options.
+  // This way unnecessary branching is avoided.
+  // The most important thing is that the non pretty printed format is optimized for
+  // as this is where the performance is most likely to matter.
```

Comment:
I agree that we should do this, but it seems like we already are generating specific functions based on the options. Is this comment outdated?

Reply:
I dont quite know, but i think it may referring todo it ahead of time, for things such as `pretty`, I could be completely wrong though.